src/rpc/reflect/openrpc_types.rs

```diff
@@ -0,0 +1,219 @@
+    pub name: String,
+    /// > REQUIRED.
+    /// > Schema that describes the content.
+    ///
+    /// > The Schema Object allows the definition of input and output data types.
+    /// > The Schema Objects MUST follow the specifications outline in the JSON Schema Specification 7 Alternatively,
+    /// > any time a Schema Object can be used, a Reference Object can be used in its place.
+    /// > This allows referencing definitions instead of defining them inline.
+    ///
+    /// > This object MAY be extended with Specification Extensions.
+    pub schema: Schema,
+    /// > Determines if the content is a required field. Default value is false.
+    #[serde(default)]
+    pub required: bool,
+}
+
+/// > REQUIRED.
+/// > A list of parameters that are applicable for this method.
+/// > The list MUST NOT include duplicated parameters and therefore require name to be unique.
+/// > The list can use the Reference Object to link to parameters that are defined by the Content Descriptor Object.
+/// > All optional params (content descriptor objects with "required": false) MUST be positioned after all required params in the list.
+#[derive(Debug, Clone, PartialEq, Serialize, Default)]
+#[serde(transparent)]
+pub struct Params {
+    inner: Vec<ContentDescriptor>,
+}
+
+#[derive(Debug, Clone, Error)]
+#[error("{}", .0)]
+pub struct ParamListError(String);
+
+impl Params {
+    pub fn new(
+        params: impl IntoIterator<Item = ContentDescriptor>,
+    ) -> Result<Self, ParamListError> {
+        let params = params.into_iter().collect::<Vec<_>>();
+        let duplicates = params
+            .iter()
+            .map(|it| it.name.as_str())
+            .duplicates()
+            .collect::<Vec<_>>();
+        if !duplicates.is_empty() {
+            return Err(ParamListError(format!(
+                "The following parameter names are duplicated: [{}]",
+                duplicates.join(", ")
+            )));
+        }
+        if let Some((first_opt_ix, first_opt_param)) =
+            params.iter().enumerate().find(|(_, it)| !it.required)
+        {
+            let late_mandatory_params = params
+                .iter()
+                .enumerate()
+                .filter(|(ix, it)| it.required && *ix > first_opt_ix)
+                .map(|(_, it)| it.name.clone())
+                .collect::<Vec<_>>();
+
+            if !late_mandatory_params.is_empty() {
+                return Err(ParamListError(format!(
+                    "Mandatory parameters may not follow optional parameters, \
+                            but the optional parameter {} is followed by [{}]",
+                    first_opt_param.name,
+                    late_mandatory_params.join(", ")
+                )));
+            }
+        };
+        Ok(Self { inner: params })
+    }
+}
+
+validated_vec!(Params => ContentDescriptor);
+
+macro_rules! validated_vec {
```

Comment:
nit: why not a derive macro? Seems more logical. I understand that you need to supply a param, but that's easy enough using attributes.

Reply:
I guess the need to define it in a different module is a pain.
Let's make sure that all the macros are documented then, it would be easier to maintain and review.

Reply:
derive macros would need to be a different crate. I'm not against writing one, but this is much simpler